shared/Microsoft.AspNetCore.HttpSys.Sources/RequestProcessing/NativeRequestContext.cs

```diff
@@ -4,33 +4,49 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Net.Sockets;
 using System.Runtime.InteropServices;
-using System.Security.Claims;
 using System.Security.Principal;
 using Microsoft.Extensions.Primitives;
 
-namespace Microsoft.AspNetCore.Server.HttpSys
+namespace Microsoft.AspNetCore.HttpSys.Internal
 {
     internal unsafe class NativeRequestContext : IDisposable
     {
-        private const int DefaultBufferSize = 4096;
         private const int AlignmentPadding = 8;
-        private HttpApi.HTTP_REQUEST* _nativeRequest;
-        private IntPtr _originalBufferAddress;
+        private HttpApiTypes.HTTP_REQUEST* _nativeRequest;
         private byte[] _backingBuffer;
+        private IntPtr _originalBufferAddress;
         private int _bufferAlignment;
         private SafeNativeOverlapped _nativeOverlapped;
-        private AsyncAcceptContext _acceptResult;
 
-        internal NativeRequestContext(AsyncAcceptContext result)
+        // To be used by HttpSys
+        internal NativeRequestContext(SafeNativeOverlapped nativeOverlapped,
+            int bufferAlignment,
+            HttpApiTypes.HTTP_REQUEST* nativeRequest,
+            byte[] backingBuffer,
+            ulong? requestId = null)
+        {
+            _nativeOverlapped = nativeOverlapped;
+            _bufferAlignment = bufferAlignment;
+            _nativeRequest = nativeRequest;
+            _backingBuffer = backingBuffer;
+            if (requestId != null)
+            {
+                RequestId = requestId.Value;
+            }
+        }
+
+        // To be used by IIS Integration.
+        internal NativeRequestContext(HttpApiTypes.HTTP_REQUEST* request)
         {
-            _acceptResult = result;
-            AllocateNativeRequest();
+            _nativeRequest = request;
```

Comment:
needs a flag like `_perminatelyPinned` so all the lazy methods using `_backingBuffer` can have a non-lazy version for IIS.